teps/0032-tekton-notifications.md

```diff
@@ -0,0 +1,167 @@
+different platforms.
+
+### Non-Goals
+
+- Build a machine to machine communication protocols
+- Build pre/post execution hook for steps, tasks or pipelines
+- Build a catalog of integration with different notification, chat or email systems
+- Build a system to send large amounts of data produced by a task or pipeline
+- Write a controller for cloud events (even if that could be part of the solution).
+  See [tektoncd/pipeline#2944](https://github.com/tektoncd/pipeline/issues/2944) for
+  further context.
+
+### Use Cases (optional)
+
+- Notify individuals or groups about the failure of a pipeline
+- Notify individuals or groups about the completion of a long running task
+- Notify individuals or groups about the start or a pipeline that was triggered as
+  a consequence of an automated event (cron, pull request, manual approval)
+
+## Requirements
+
+This section is a stub and subject to change.
+
+Once notifications are implemented, we do not want to force Tekton users to modify
+all their existing tasks or pipelines to enable notifications for them:
+
+- Allow using catalog tasks as is
+- Allow notifications for pipelines with no change to the pipeline itself
+
+For installations that run large number of task and pipelines, the configuration
+of notification should be designed to be scalable and easy to maintain.
+
+- Allow selecting pipelines and tasks to be notified about via selectors
+- Allow selecting events to be notified about
+- Allow pipeline and task authors to re-use of notification "rules"
+
+## Proposal
+
+This section is a stub and subject to change.
+
+The proposal is to introduce notification policies and templates.
+
+Notification policies define condition under which we want to trigger a notification;
+they would allow users to:
+
+- filter sources such using rules (using CEL?)
+  - kind: task, pipeline
+  - metadata: name, labels, namespace, annotations(?)
+- filter events such as start, run, success, fail (using CEL or syntax sugar on top)
+
+Notification templates define how a notification is sent. They would allow user to:
+
+- define the task / pipeline to be executed
+- bind an event type with the inputs to the task / param
+
+Both notification policies and templates could be defined a dedicated CRD.
+The policy controller would be responsible to provision `trigger` resources by
+combining the information from the policy and the templates.
+
+Which notification template is triggered is decided based on runtime information passed to the `TaskRun`
+or `PipelineRun`, or can be configured on Tekton for each namespace and added to resources
+as an annotation.
+
+### User Experience (optional)
+
+Notification targeted to humans is the main goal of this TEP because I would like to make
+sure that requirements for machine targeted notification do not pollute the user experience
+of tekton native notifications.
+
+User experience is very important for this proposal. Notification should be easy to use
```

Comment:
Do we want to mention anything here about how auth/secrets work (e.g. chat app webhook URL)? I think these could just be Kubernetes secrets, right?

Reply:
My thought was to leave this aspect as a responsibility of the task that sends the notification. Usually they have a reference to a k8s secret where to pull the required secrets from.

We could consider surfacing that configuration into the notification template for ease of use. The main doubt I have is how to provide a standard format for that, since different types of notifications will need different type of credentials.